blackjax/tempered_smc.py

```diff
@@ -0,0 +1,116 @@
+"""Public API for the Tempered SMC Kernel"""
+from typing import Callable, Dict, List, NamedTuple, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+import blackjax.inference.base as base
+from blackjax.hmc import HMCParameters
+
+Array = Union[np.ndarray, jnp.DeviceArray]
+PyTree = Union[Dict, List, Tuple]
+
+__all__ = ["new_state", "kernel"]
+
+
+class SMCState(NamedTuple):
+    """ Current state for the tempered SMC algorithm. This encodes the current set of sampled particles
+    as well as the corresponding tempering parameter :math:`\lambda`
+    """
+    particles: PyTree[Array]
+    lamda: float
+
+
+class SMCInfo(NamedTuple):
+    """Additional information on the tempered SMC step.
+
+    weights: jnp.ndarray
+        The weights after the MCMC pass.
+    """
+    weights: jnp.ndarray
+
+
+new_state = base.new_hmc_state
+
+
+def _normalize(log_weights):
+    """ Normalize log-weights into weights
+    """
+    w = jnp.exp(log_weights - jnp.max(log_weights))
+    return w / w.sum()
+
+
+def kernel(potential_fn: Callable, mcmc_kernel: Callable, new_mcmc_state: Callable, resampling_method: Callable,
+           mcmc_parameters: HMCParameters, mcmc_steps=10):
+    """Build a Tempered SMC step.
+
+    Parameters
+    ----------
+    potential_fn: Callable
+        A function that returns the potential energy of a chain at a given position.
+    mcmc_kernel: Callable
+        A callable function that creates a kernel from `potential_function` and parameters `mcmc_parameters`
+    new_mcmc_state: Callable
+        How to create a new MCMC state for the MCMC routine to start
+    resampling_method: Callable
+        A random function that resamples generated particles based of weights
+    mcmc_parameters: HMCParameters
+        The parameters to be fed to the mcmc kernel
+    mcmc_steps: int
+        Number of MCMC steps to run per SMC step
+
+    Returns
+    -------
+    A callable that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+    """
+
+    def one_step(rng_key: jnp.ndarray, state: SMCState, delta: float) -> Tuple[SMCState, SMCInfo]:
```

Comment:
It's not exactly propagated is it? lambda is, but delta is used to increment lambda only.

Reply:
But I guess this decision can come at the end of the dev work as it would be a minimal change anyway, so maybe we should keep it for later when the rest of the design is clearer? Or is the reason why you want it in the state specific to the fact you want one_step to only take key + state?

Reply:
the pb with having delta on the state is that it asks the user to manipulate the state in between two steps, essentially doing smthg like:
```python 
particles, weights, delta, lamda = state
state = State(particles, weights, new_delta, lamda)
```
which feels a bit wasteful for no reason

Reply:
Actually, could you add the instruction where you change the value of  `delta`? I need the full picture to be able to judge.